content/blog/2018-04-03-rtika-introduction.md

```diff
@@ -0,0 +1,277 @@
+
+#### Lessons Learned
+
+I never distributed a package before on repositories such as CRAN or Github, and the rOpenSci group was the right place to learn how. The reviewers used a transparent on-boarding process and taught about good documentation and coding style. They were helping create a maintainable package by following certain standards. If I stopped maintaining `rtika`, others could use their knowledge of the same standards to take over. The vast majority of time was spent on documenting the code, the introductory vignette, and continuous testing to integrate new code.
+
+##### Connecting to Tika 
+
+Since the Tika software is mature, my coding centered around connecting R to Tika. The `rtika` software is a message broker that handles inter-process communication, and there are several ways to do this: Tika server, `rJava`, or system calls. I discovered the Linux paradigm of using files and file-like paths for most messaging, paraphrased as "everything is a file" [^5]. I wanted a method that would work on Windows and OS X and be easy to install, so chose basic system calls and the file system.
+
+This worked. R sends Tika a signal to execute code using an old-fashioned command line call, telling Tika to parse a particular batch of files. R waits. Eventually, Tika sends the signal of its completion, and R can then return with results as a character vector. Surprisingly, this process may be a good option for containerized applications running Docker. In the example later in this blog post, a similar technique is used to connect to a Docker container in a few lines of code.
+
+Communication with Tika went smoothly, but after one issue with `base::system2()` was identified. That was terminating Tika's long running process. Switching to `sys::exec_wait()` or `processx::run()` solved the issue.
+
+##### The R User Interface
+
+Many in the R community make use of `magrittr` style pipelines, so those needed to work well. The Tidy Tools Manifesto makes piping a central tenant [^6], which makes code easier to read and maintain.
+
+When writing `rtika`, I created two distinct styles of user interface. The first was a lightweight R wrapper around the Tika command line, called `tika()`. The parameters and abbreviations there mirrored the Apache Tika tool. The other functions are in the common R style found in tidy tools, and run `tika()` with certain presets (e.g. `tika_html()` outputs to 'html' ). For R users, these should be more intuitive and self-documenting.
+
+##### Responding to Reviewers
+
+During the review process, I appreciated David Gohel's [^7] attention to technical details, and that Julia Silge [^8] and Noam Ross [^9] pushed me to make the documentation more focused. I ended up writing about each of the major functions in a vignette, one by one, in a methodical manner. While writing, I learned to understand Tika better.
```

Comment:
This is awesome "I ended up writing about each of the major functions in a vignette, one by one, in a methodical manner. While writing, I learned to understand Tika better."